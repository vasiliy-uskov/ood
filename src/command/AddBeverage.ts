import {Capuccino} from "../beverage/Capuccino";
import {Espresso} from "../beverage/Espresso";
import {Latte} from "../beverage/Latte";
import {Milkshake} from "../beverage/Milkshake";
import {Tea} from "../beverage/Tea";
import {IBeverage} from "../beverage/IBeverage";
import {Order} from "../Order";
import {ICommand} from "./ICommand";

const BeveragesTypes: {[key:string]: () => IBeverage} = {
	'capuccino': () => new Capuccino(),
	'espresso': () => new Espresso(),
	'latte': () => new Latte(),
	'milkshake': () => new Milkshake(),
	'tea': () => new Tea(),
};

class AddBeverage implements ICommand {
	constructor(order: Order) {
		this._order = order;
	}

	execute(command: string) {
		const matchResult = command.toLowerCase().match(/make (\w+)/);
		if (matchResult) {
			const type = matchResult[1];
			if (BeveragesTypes[type]) {
				this._order.addBeverage(BeveragesTypes[type]());
				return;
			}
		}
		throw new Error(`Wrong command!\nTry: ${this.help()}`)
	}

	help(): string {
		return `make <${this._getBeverageTypes()}>\n`
	}

	is(command: string): boolean {
		return !!command.toLowerCase().match(/make.*/)
	}

	private _getBeverageTypes(): string {
		return Object.keys(BeveragesTypes).join('|');
	}

	private readonly _order: Order;
}

export {AddBeverage}