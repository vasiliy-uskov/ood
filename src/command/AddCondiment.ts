import {Order} from "../Order";
import {ICommand} from "./ICommand";
import {
	ChocolateCrumbsCreator,
	ChocolateSlicesCreator,
	CinnamonCreator,
	CoconutFlakesCreator,
	CreamCreator,
	IceCubesCreator,
	ICondimentCreator,
	LemonCreator, LiquorCreator,
	SyrupCreator
} from "./condiments/CondimentsCreators";

const condimentCreators: Array<Readonly<ICondimentCreator>> = [
	new ChocolateCrumbsCreator(),
	new CinnamonCreator(),
	new CoconutFlakesCreator(),
	new LemonCreator(),
	new IceCubesCreator(),
	new SyrupCreator(),
	new LiquorCreator(),
	new ChocolateSlicesCreator(),
	new CreamCreator(),
];

class AddCondiment implements ICommand {
	constructor(order: Order) {
		this._order = order;
	}

	execute(command: string): void {
		const matchResult = command.toLowerCase().match(/add (([a-z]|\s)+)(: (.*))?$/);
		if (matchResult) {
			const condiment = matchResult[1];
			const args = matchResult[3];
			for (const creator of condimentCreators) {
				if (creator.is(condiment)) {
					this._order.addCondiment(creator.create(args));
					return
				}
			}
		}
		throw new Error(`Wrong command! Try: \n${this.help()}`)
	}

	help(): string {
		return `add <
${condimentCreators
			.map(creator => `	${creator.help()}`)
			.join('|\n')}
>\n`;
	}

	is(command: string): boolean {
		return !!command.toLowerCase().match(/add.*/)
	}

	private readonly _order: Order;
}


export {AddCondiment}