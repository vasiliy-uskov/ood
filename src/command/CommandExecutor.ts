import {ICommand} from "./ICommand";

export class CommandExecutor {
	constructor(commands: Array<ICommand>) {
		this._commands = commands.slice();
	}

	execute(commandStr: string) {
		const command = this._commands.find((command) => command.is(commandStr));
		if (!command)
		{
			console.log('Undefined command.');
		}
		else
		{
			try
			{
				command.execute(commandStr);
			}
			catch (e) {
				console.log(e);
			}
		}
	}

	private readonly _commands: Array<ICommand>;
}
