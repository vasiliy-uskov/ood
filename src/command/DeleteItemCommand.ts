import {IEditor} from "../IEditor";
import {WrongCommandError} from "./WrongCommandError";

export class DeleteItemCommand {
	constructor(editor: IEditor) {
		this._editor = editor;
	}

	is(command: string): boolean {
		return !!command.match(/^DeleteItem.*/);
	}

	execute(command: string) {
		const res = command.match(/^DeleteItem\s(\d+)$/);
		if (!res)
		{
			throw new WrongCommandError(this.help());
		}
		const position = parseInt(res[1]);
		this._editor.deleteItem(position);
	}

	help(): string {
		return 'DeleteItem <позиция> <текст параграфа>';
	}
	private _editor: IEditor;
}