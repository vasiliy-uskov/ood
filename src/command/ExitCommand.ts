export class ExitCommand {
	is(command: string): boolean {
		return !!command.match(/^Exit$/);
	}

	execute() {
		process.exit(0);
	}

	help(): string {
		return 'Exit';
	}
}

