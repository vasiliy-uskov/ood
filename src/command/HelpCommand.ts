export class HelpCommand {
	constructor(commands: Array<{help: () => string}>) {
		this._commands = commands;
	}

	is(command: string): boolean {
		return !!command.match(/^Help$/);
	}

	execute() {
		console.log(`Доступныу команды:\n\t${this._commands.map(command => command.help()).join('\n\t')}\n`);
	}

	help(): string {
		return 'Help';
	}
	private _commands: Array<{help: () => string}>;
}

