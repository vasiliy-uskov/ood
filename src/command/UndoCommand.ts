import {IEditor} from "../IEditor";

export class UndoCommand {
	constructor(editor: IEditor) {
		this._editor = editor;
	}

	is(command: string): boolean {
		return !!command.match(/^Undo$/);
	}

	execute() {
		this._editor.undo();
	}

	help(): string {
		return 'Undo';
	}
	private _editor: IEditor;
}