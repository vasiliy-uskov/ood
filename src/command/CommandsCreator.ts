import {IEditor} from "../IEditor";
import {ICommand} from "./ICommand";
import {DeleteItemCommand} from "./DeleteItemCommand";
import {InsertImageCommand} from "./InsertImageCommand";
import {InsertParagraphCommand} from "./InsertParagraphCommand";
import {ReplaceParagraphTextCommand} from "./ReplaceParagraphTextCommand";
import {ResizeImageCommand} from "./ResizeImageCommand";
import {SaveDocumentCommand} from "./SaveDocumentCommand";
import {UndoCommand} from "./UndoCommand";
import {RedoCommand} from "./RedoCommand";
import {ListCommand} from "./ListCommand";
import {SetDocumentTitleCommand} from "./SetDocumentTitleCommand";
import {ExitCommand} from "./ExitCommand";
import {HelpCommand} from "./HelpCommand";
import {DocumentLogger} from "../view/DocumentLogger";
import {HTMLDocumentSaver} from "../view/HTMLDocumentSaver";

export function createCommands(editor: IEditor): Array<ICommand> {
	const commands: Array<ICommand> = [
		new DeleteItemCommand(editor),
		new InsertImageCommand(editor),
		new InsertParagraphCommand(editor),
		new ReplaceParagraphTextCommand(editor),
		new ResizeImageCommand(editor),
		new SaveDocumentCommand(editor, HTMLDocumentSaver.save),
		new UndoCommand(editor),
		new RedoCommand(editor),
		new ListCommand(editor, DocumentLogger.log),
		new SetDocumentTitleCommand(editor),
		new ExitCommand(),
	];
	const helpCommand = new HelpCommand(commands);
	commands.push(helpCommand);
	return  commands;
}