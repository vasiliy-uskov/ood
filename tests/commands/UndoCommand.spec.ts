import {UndoCommand} from "../../src/command/UndoCommand";
import {createEditor} from "../mocks/DocumentData";

it('return help', () => {
	const command = new UndoCommand(createEditor());
	expect(command.help()).toMatchSnapshot();
});

describe('is', () => {
	it('return true if command match Undo', () => {
		const command = new UndoCommand(createEditor());
		expect(command.is('Undo')).toBe(true);
		expect(command.is('Undo ')).toBe(false);
		expect(command.is('Undo.')).toBe(false);
		expect(command.is(' Undo')).toBe(false);
		expect(command.is('Red')).toBe(false);
		expect(command.is('')).toBe(false);
	});
});
describe('execute', () => {
	it('call undo method without any arguments', () => {
		const editor = createEditor();
		const command = new UndoCommand(editor);
		expect(editor.undo).not.toBeCalled();
		command.execute();
		expect(editor.undo).toBeCalledTimes(1);
		expect(editor.undo).toBeCalledWith();
	});
});