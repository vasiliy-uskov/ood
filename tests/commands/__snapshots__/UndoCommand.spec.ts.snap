// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`return help 1`] = `"Undo"`;
